sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreCoordinator.scala

```diff
@@ -150,6 +172,25 @@ private class StateStoreCoordinator(override val rpcEnv: RpcEnv)
         storeIdsToRemove.mkString(", "))
       context.reply(true)
 
+    case ValidateSchema(providerId, keySchema, valueSchema, checkEnabled) =>
+      // normalize partition ID to validate only once for one state operator
+      val newProviderId = StateStoreProviderId.withNoPartitionInformation(providerId)
+
+      val result = schemaValidated.getOrElseUpdate(newProviderId, {
+        val checker = new StateSchemaCompatibilityChecker(newProviderId, hadoopConf)
+
+        // regardless of configuration, we check compatibility to at least write schema file
+        // if necessary
+        val ret = Try(checker.check(keySchema, valueSchema)).toEither.fold(Some(_), _ => None)
```

Comment:
Roughly remember we don't recommend using `Try` in Spark?

Reply:
https://github.com/databricks/scala-style-guide#exception

Do not use `Try` is applied only to public API, that sounds to me as do not require callers to deal with `Try`. You can also find there's usage of Try in Spark codebase.

Reply:
I'm wondering if we have a better way to do this. We don't encourage to read remote files in RPC threads. If reading the remote file takes several minutes, it might take down the whole Spark driver.

Reply:
https://github.com/apache/spark/pull/24173/files#r534772017

Let's just check it without sending RPC. We'll need some trick - hope that's acceptable.